### PR TITLE
Delete device functionality

### LINKED ISSUE #287: There is no way to remove old devices
Currently, we can add new devices via `eas device:create`. However, there is no way to remove devices.

### PR DESCRIPTION
**Pull request by kbrandwijk**
# Checklist

- [X] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Resolves #287 

# How

- [X] Add `device:delete` command for device selection
- [X] Delete device using existing gql api
- [X] Update adhoc build provision logic to take deleted devices into account - not needed

# Test Plan

- [X] Manually tested deleting, and subsequent `device:list` not displaying the device anymore. Also tested how this affects the adhoc provisioning after removing a device that was previously provisionined for adhoc deployment on a project.

### CODE REVIEW

File: packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts
Comment (dsokal) [reply packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts:35]: Generally, I think it'd be better to let users delete multiple devices at once.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/device/delete.ts:41]:
```suggestion
        if (teams.length > 0) {
```
- It's always better to be explicit instead of depending on implicit conversion.
- Yes, I know this is copy-pasted.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/device/delete.ts:23]: I'd be nice if this command accepted the device udid, so you could remove a device like this:
`eas device:delete --apple-team-id TEAM_ID --udid UDID`

File: packages/eas-cli/src/commands/device/delete.ts
Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:41]: I'll adjust it in the original location too.

File: packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts
Comment (kbrandwijk) [reply packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts:35]: Just to be clear, do you want the behavior adjusted or the question adjusted?

Comment (dsokal) [reply packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts:35]: I'd prefer to change the behavior (so it's possible to delete multiple devices with a single `eas device:delete` run). If you think this doesn't make sense then we need to change the question.

Comment (kbrandwijk) [reply packages/eas-cli/src/credentials/ios/actions/DeviceUtils.ts:35]: Multi-select implemented

File: packages/eas-cli/src/commands/device/delete.ts
Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:24]: Parameter implemented, including multi-select

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:30]: run parse once

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:109]: you are not returning anything, use forEach or for loop

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:111]: use Log.newline instead of `\n`

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:121]:
```suggestion
            `You are about to remove the Apple device ${
```

Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:121]: No, it adds an -s for multiple.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:139]:
```suggestion
              await Promise.all(chosenDevices.map(chosenDevice => {
                 return AppleDeviceMutation.deleteAppleDeviceAsync(chosenDevice.id);
              }));
```
you are not awaiting for those values, forEach works only with synchronous methods (alternatively you could use for loop if it has to be called one by one)

plus nit pick: Irecomend moving `removalSpinner.succeed` in the try  block at the end

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:163]: the same as above
- forEach is not awaiting correctlly
- spinner.succeed in try block

File: packages/eas-cli/src/commands/device/delete.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/device/delete.ts:30]: This function is 130 LOC. That's a lot. Please break it down.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/device/delete.ts:154]: I think `: RequestContext` is not necessary.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (dsokal) [inline packages/eas-cli/src/commands/device/delete.ts:169]: Maybe we should be explicit here. We actually don't **remove** the devices, we **disable** them.

Comment (dsokal) [reply packages/eas-cli/src/commands/device/delete.ts:169]: Change that in other logs as well.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:139]: Reading up on this, I will actually turn this into a normal for loop, as I cannot control the degree of parallelism and I don't want to run into weird errors if a user decides to remove 20 devices at the same time.

File: packages/eas-cli/src/commands/device/delete.ts
Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:30]: `list` is 100 LOC. What's the target value?

Comment (dsokal) [reply packages/eas-cli/src/commands/device/delete.ts:30]: Basically, I think we should have a lint rule for that. It's hard to add that in right now as it'd require a lot of refactoring. I'd say a rule of thumb is that a function should fit a reasonable display (without scrolling).

Comment (kbrandwijk) [reply packages/eas-cli/src/commands/device/delete.ts:30]: Works for me, I'll make a note to include that rule for myself whenever I need to 'touch' existing code as well.

**Review by dsokal — APPROVED** [review]
LGTM, let's wait for a review from @wkozyra95

File: packages/eas-cli/src/commands/device/delete.ts
Comment (wkozyra95) [inline packages/eas-cli/src/commands/device/delete.ts:149]:
```suggestion
        formatDevice(device, {
          appleTeamName,
          appleTeamIdentifier: appleTeamIdentifier!,
        })
```

**Review by wkozyra95 — APPROVED** [review]

**Pull request merged**
